raster/src/main/scala/geotrellis/raster/costdistance/CostDistance.scala

```diff
@@ -16,236 +16,190 @@
+    val cols = frictionTile.cols
+    val rows = frictionTile.rows
+    val costTile = generateEmptyCostTile(cols, rows)
+    val q: Q = generateEmptyQueue(cols, rows)
+
+    points.foreach({ case (col, row) =>
+      val entry = (col, row, frictionTile.getDouble(col, row), 0.0)
+      q.add(entry)
+    })
+
+    compute(frictionTile, costTile, maxCost, resolution, q, nop)
   }
 
-  val dirs: Array[Dir] = Array(
-    (-1, -1), ( 0, -1), ( 1, -1),
-    (-1, 0),          ( 1, 0),
-    (-1, 1), ( 0, 1), ( 1, 1)).map { case (c, r) => new Dir(c, r) }
-
-
   /**
-    * Input:
-    * (c, r) => Source cell
-    * (dc, dr) => Delta (direction)
-    * cost => Cost raster
-    * d => C - D output raster
+    * Compute a cost tile.
     *
-    * Output:
-    * List((c, r)) <- list of cells set
+    * @param  frictionTile    The friction tile
+    * @param  costTile        The tile that will contain the costs
+    * @param  maxCost         The maximum cost before pruning a path (in units of "seconds")
+    * @param  resolution      The resolution of the tiles (in units of "meters per pixel")
+    * @param  q               A priority queue of Cost objects (a.k.a. candidate paths)
+    * @param  edgeCallback    Called when a pixel on the edge of the tile is updated
     */
-  private def calcCostCell(c: Int, r: Int, dir: Dir, cost: Tile, d: DoubleArrayTile) = {
-    val cr = dir(c, r)
-
-    if (isValid(cr._1, cr._2, cost)) {
-      val prev = d.getDouble(cr._1, cr._2)
-      if (prev == 0.0) { // This is a source cell, don't override and shortcircuit early
-        None
-      } else {
-        val source = d.getDouble(c, r)
-
-        // Previous value could be NODATA
-        val prevCost = if (isNoData(prev)) java.lang.Double.MAX_VALUE else prev
-
-        var curMinCost = Double.MaxValue
-
-        // Simple solution
-        val baseCostOpt = calcCost(c, r, dir, cost)
-        if (baseCostOpt.isDefined) {
-          curMinCost = source + baseCostOpt.get
-        }
-
-        // Alternative solutions (going around the corner)
-        // Generally you can check diags directly:
-        // +---+---+---+
-        // | a | b | c |
-        // +---+---+---+
-        // | d | e | f |
-        // +---+---+---+
-        // | g | h | i |
-        // +---+---+---+
-        //
-        // For instance, "eg" can be computed directly
-        // but it turns out "eg" can be more expensive
-        // than "edg" or "ehg" so we compute those right
-        // now just in case
-        val cornerOffsets = dir.cornerOffsets
-        val l = cornerOffsets.length
-        var z = 0
-        while(z < l) {
-          val p = cornerOffsets(z)
-          val c1 = p._1 + c
-          val r1 = p._2 + r
-          val cost1 = calcCost(c, r, c1, r1, cost)
-          if (cost1.isDefined) {
-            val cost2 = calcCost(c1, r1, dir(c, r), cost)
-            if (cost2.isDefined) {
-              curMinCost = math.min(curMinCost, source + cost1.get + cost2.get)
+  def compute(
```

Comment:
I have not run any benchmarks, but (informally) I did not detect any noticeable speed difference.

That having been said, it is likely that there is one, because the changes that I made were very self-consciously **deoptimizations**.

That was necessary because the previous algorithm seemed to avoid putting elements into the priority queue whenever it could.  That was laudable in the single-threaded case, but in order to maintain coherence in the case were points need to be transferred from some adjacent tile to the present one, I think that it makes sense to have one and only one place for points to enter (to ensure that they are being treated exactly the same as points that originated in the present tile).

After getting the basics working, I then took another pass an re-added some mild optimizations.  Generally speaking, I am pretty comfortable with were the single-threaded version is.